### PR TITLE
Add check for 'HTMLElement' type

### LINKED ISSUE #78: Circular structure bug when referencing the $el in a different variable
Referencing the $el in another object attribute will cause a bug when the devtool stringifies the object.

```javascript
var component = function() {
  return(
    {
      msg: 'Hello Alpine.js',
      init() {
        this.div = this.$el 
      }
    }
   )
}
```
See this page here with the console opened: https://cdpn.io/pomartel/debug/abmbyNg/xJAjOeOZonJk
The Codepen source here: https://codepen.io/pomartel/pen/abmbyNg

### PR DESCRIPTION
**Pull request by KevinBatdorf**
Fixes #78 

Serializes HTMLElement types and changes the logic on what is readOnly and what readOnly can do.

`readOnly` only affects editing now, not opening/closing

### CODE REVIEW

File: packages/shell-chrome/src/backend.js [hunk 1]
```diff
@@ -144,6 +144,9 @@ function discoverComponents(isThroughMutation = false) {
 
         const data = Object.entries(rootEl.__x.getUnobservedData()).reduce((acc, [key, value]) => {
             const type = typeof value
+            if (value instanceof HTMLElement) {
```
Comment (HugoDF) [inline packages/shell-chrome/src/backend.js:147]: funny, I don't remember what we use the `type` field for

I was thinking we should probably extract a `replaceUnsupportedType(value)` function that does

```js
const replaceUnsupportedType = (value) => {
  if (typeof value === 'function') return 'function'
  if (value instanceof HTMLElement) return 'HTMLElement'
  return value
}
```

**Comment by HugoDF** [comment]
Re tests let's chat in https://github.com/alpine-collective/alpinejs-devtools/issues/81

File: packages/shell-chrome/src/backend.js (hunk 1 above)
Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:147]: `reaplceUnsupportedType` vs `replaceNonSerializableTypes` 🤔

**Comment by HugoDF** [comment]
Btw the project can now be tested with Cypress when changing `backend.js` or the Alpine.js Panel app.

I've stuck the test for this PR on the todo list at #86 but we can add it here as well.

File: packages/shell-chrome/src/backend.js (hunk 1 above)
Comment (KevinBatdorf) [reply packages/shell-chrome/src/backend.js:166]: You think it should go in the utils.js file from #79 and be a jest test too (can do cypress too)?

Are there any others to check for? maybe setters.

Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:166]: So it turns out we use the `type` field for showing things as "readOnly" (so no checkbox or edit buttons)

I don't really mind if we create the new function here or in that utils file, we can test with Jest either way

Would be nice to cover "function and HTMLElement are displayed and not editable" with a Cypress test

Comment (KevinBatdorf) [reply packages/shell-chrome/src/backend.js:166]: Okay cool. So probably need to set the type, or maybe refactor that out somehow. I'll look into that later.

**Comment by HugoDF** [comment]
Hmmm on second thought, it looks like we need to rethink the way we pass "readonly" attributes to the panel, see the screenshots, `el` was set as `el: $el` in x-data, it's editable

Screenshot 2020-12-01 at 18 42 31
Screenshot 2020-12-01 at 18 42 36

potentially instead of passing `type` we should also pass a `readOnly: true/false` field here:

https://github.com/alpine-collective/alpinejs-devtools/blob/29530784934bd40c70c152f42020b65fd3c30432/packages/shell-chrome/src/backend.js#L169-L172

instead of doing the whole "if type === 'function'" in the place that receives the message https://github.com/alpine-collective/alpinejs-devtools/blob/78466dd48ea16ce1f68a3478564f66e179e5d1c3/packages/shell-chrome/src/utils.js#L107

**Comment by KevinBatdorf** [comment]
I think this turned out well. Can you take a look when you have time? I can add some tests later this week. I feel I have a pretty decent grasp as to how the codebase works now :)

devtools

Also, I noticed in the simulator the arrows aren't showing. I think we just need to copy over some CSS that's inlined. I wasn't sure so left it be. We can do it in another branch easily enough though.

File: packages/shell-chrome/src/backend.js [hunk 2]
```diff
@@ -86,7 +86,8 @@ function handleMessages(e) {
                         data[attributeSequence] = attributeValue
                     }
 
-                    component.__x.$el.setAttribute('x-data', JSON.stringify(data))
+                    component.__x.$data = data
+                    component.__x.updateElements(component)
```
Comment (KevinBatdorf) [inline packages/shell-chrome/src/backend.js:90]: This is possibly a big change so I want to point it out.

Also, on the Inline devtools I try to only update the specific property which I think is much faster, so maybe something to tackle here on another day
https://github.com/KevinBatdorf/alpine-inline-devtools/blob/master/src/Viewer.js#L69

Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:90]: I think this will be fine, was not a fan of overwriting `x-data`

File: packages/shell-chrome/src/backend.js [hunk 3]
```diff
@@ -127,10 +128,9 @@ function discoverComponents(isThroughMutation = false) {
         }
 
         const data = Object.entries(rootEl.__x.getUnobservedData()).reduce((acc, [key, value]) => {
-            const type = typeof value
             acc[key] = {
-                value: type === 'function' ? 'function' : value,
-                type,
+                value: value instanceof HTMLElement ? serializeHTMLElement(value) : value,
+                type: value instanceof HTMLElement ? 'HTMLElement' : typeof value,
```
Comment (HugoDF) [inline packages/shell-chrome/src/backend.js:133]: This is a bit odd since HTMLElement isn't a real JS type

Maybe we should add a comment to look at this line wherever we're processing the message being built here

File: packages/shell-chrome/src/utils.js [hunk 4]
```diff
@@ -87,24 +87,52 @@ export function convertInputDataToType(inputType, value) {
     }
 }
 
+function getAttributeValue(value, type) {
```
Comment (HugoDF) [inline packages/shell-chrome/src/utils.js:90]: getDisplayName?

Let's leave this for another PR, we use "attribute" everywhere so makes sense to keep it consistent for now

**Review by HugoDF — COMMENTED** [review]
Nice one on the element name, just a few questions.

Have you tried it out with `els: [$el]`? Just wondering if it magically works otherwise I think this is good to merge and we can create a separate issue to deal with HTML elements in nested object/arrays

File: packages/shell-chrome/src/backend.js [hunk 5]
```diff
@@ -127,10 +128,9 @@ function discoverComponents(isThroughMutation = false) {
         }
 
         const data = Object.entries(rootEl.__x.getUnobservedData()).reduce((acc, [key, value]) => {
-            const type = typeof value
             acc[key] = {
-                value: type === 'function' ? 'function' : value,
-                type,
+                value: value instanceof HTMLElement ? serializeHTMLElement(value) : value,
```
Comment (HugoDF) [inline packages/shell-chrome/src/backend.js:132]: we've lost the `'function'` replace here, since we're building a message payload, it would be better to do the function -> `'function'` replace before sending the message (since it will keep the payload smaller)

```suggestion
                value: value instanceof HTMLElement ? serializeHTMLElement(value) : typeof value === 'function' ? 'function' : value,
```

**Comment by HugoDF** [comment]
> Also, I noticed in the simulator the arrows aren't showing. I think we just need to copy over some CSS that's inlined. I wasn't sure so left it be. We can do it in another branch easily enough though.

Fixed this in #95

File: packages/shell-chrome/src/backend.js (hunk 5 above)
Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:132]: If we apply this change, I think we're ready to merge 😄 we can do tests in another PR

File: packages/shell-chrome/src/backend.js (hunk 2 above)
Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:90]: On second thought, would it be possible to make this a separate PR/issue?

Would like to release this `HTMLElement` fix without changing the way we update user's component state.

Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:90]: I like the idea of only updating a single field, could be something like

```js
if (attributeSequence.includes('.')) {
  set(component.__x.$data, attributeSequence, attributeValue)
  component.__x.updateElements(component) // force re-render on Array/Object update?
} else {
  component.__x.$data[attributeSequence] = attributeValue
}
```

but let's do that as a separate performance thing

Comment (HugoDF) [reply packages/shell-chrome/src/backend.js:90]: ok I'll remove this and we can have another look in https://github.com/alpine-collective/alpinejs-devtools/issues/97
